diff --git a/pkg/profile/stacktraces.go b/pkg/profile/stacktraces.go
index c58a7ff..622c9a1 100644
--- a/pkg/profile/stacktraces.go
+++ b/pkg/profile/stacktraces.go
@@ -31,7 +31,7 @@ func (p *Profiler) getStackTraceByID(stackTraces *bpf.BPFMap, stackID uint32) (*
 // Symbolization is supported for non-stripped ELF executable binaries, because the .symtab
 // ELF section is looked up.
 func (p *Profiler) getHumanReadableStackTrace(stackTrace *StackTrace) []string {
-	symbols := make([]string, 0)
+	symbols := make([]string, len(stackTrace))
 
 	for _, ip := range stackTrace {
 		if ip == 0 {
