diff --git a/runtime/src/bpftime_shm_internal.cpp b/runtime/src/bpftime_shm_internal.cpp
index 856bb8a..d42bc13 100644
--- a/runtime/src/bpftime_shm_internal.cpp
+++ b/runtime/src/bpftime_shm_internal.cpp
@@ -72,7 +72,8 @@ static __attribute__((destructor(65535))) void __destruct_shm()
 {
 	// This usually indicates that the living shared memory object is used
 	// by an agent instance
-	if (bpftime::shm_holder.global_shared_memory.get_open_type() ==
+	if (global_shm_initialized &&
+	    bpftime::shm_holder.global_shared_memory.get_open_type() ==
 	    bpftime::shm_open_type::SHM_OPEN_ONLY) {
 		// Try our best to remove the current pid from alive agent's set
 		int self_pid = getpid();
@@ -1103,7 +1104,9 @@ void bpftime_shm::add_pid_into_alive_agent_set(int pid)
 }
 void bpftime_shm::remove_pid_from_alive_agent_set(int pid)
 {
-	injected_pids->erase(pid);
+	if (injected_pids != nullptr) {
+		injected_pids->erase(pid);
+	}
 }
 void bpftime_shm::iterate_all_pids_in_alive_agent_set(
 	std::function<void(int)> &&cb)
